src/Settings/Builders/RunTest.hs

```diff
@@ -0,0 +1,113 @@
+module Settings.Builders.RunTest (runTestBuilderArgs) where
+
+import Hadrian.Utilities
+import Hadrian.Haskell.Cabal
+
+import Flavour
+import Rules.Test
+import Settings.Builders.Common
+import Settings.Builders.Ghc
+
+-- Arguments to send to the runtest.py script.
+runTestBuilderArgs :: Args
+runTestBuilderArgs = mconcat
+    [ builder RunTest ? do
+        pkgs     <- expr $ stagePackages Stage1
+        libTests <- expr $ filterM doesDirectoryExist $ concat
+                [ [ pkgPath pkg -/- "tests", pkgPath pkg -/- "tests-ghc" ]
+                | pkg <- pkgs, isLibrary pkg, pkg /= rts, pkg /= libffi ]
+
+        debugged          <- ghcDebugged <$> expr flavour
+
+        withNativeCodeGen <- expr ghcWithNativeCodeGen
+        withInterpreter   <- expr ghcWithInterpreter
+        unregisterised    <- expr $ flag GhcUnregisterised
+
+        withSMP           <- expr ghcWithSMP
+
+        windows  <- expr windowsHost
+        darwin   <- expr osxHost
+
+        threads  <- shakeThreads <$> expr getShakeOptions
+        verbose  <- shakeVerbosity <$> expr getShakeOptions
+
+        top      <- expr topDirectory
+        compiler <- expr $ builderPath $ Ghc CompileHs Stage2
+        ghcPkg   <- expr $ builderPath $ GhcPkg Update Stage1
+        haddock  <- expr $ builderPath $ Haddock BuildPackage
+        hp2ps    <- expr $ builderPath $ Hp2Ps
+        hpc      <- expr $ builderPath $ Hpc
+
+        let ifMinGhcVer ver opt = do v <- expr ghcCanonVersion
+                                     if ver <= v then pure opt
+                                                 else pure ""
```

Comment:
Isn't this simply `(ver <= v) ? pure opt`?

Reply:
Now I move this port to `src/Rules/Tests.hs`. I still use `if ... then ... else` because operator `?` is only available for `Expr` monad, can't be used in `Action` monad.